core/shared/src/main/scala/fs2/Scan.scala

```diff
@@ -0,0 +1,242 @@
+      onComplete_
+    )
+
+  /** Transforms the state type. */
+  def imapState[S2](g: S => S2)(f: S2 => S): Scan[S2, I, O] =
+    Scan[S2, I, O](g(initial))(
+      { (s2, i) =>
+        val (s3, os) = transform(f(s2), i)
+        (g(s3), os)
+      },
+      AndThen(f).andThen(onComplete_)
+    )
+
+  /** Returns a new scan with transformed input and output types.
+    *
+    * Upon receiving an `I2`, `get` is invoked and the result is fed to the
+    * original scan. For each output value, `set` is invoked with the original
+    * `I2` input and the computed `O`, yielding a new output of type `O2`.
+    */
+  def lens[I2, O2](get: I2 => I, set: (I2, O) => O2): Scan[S, I2, O2] =
+    Scan[S, I2, O2](initial)(
+      { (s, i2) =>
+        val (s2, os) = transform(s, get(i2))
+        (s2, os.map(s => set(i2, s)))
+      },
+      _ => Chunk.empty
+    )
+
+  /** Returns a scan that inputs/outputs pairs of elements, with `I` and `O` in the first element of the pair. */
+  def first[A]: Scan[S, (I, A), (O, A)] =
+    lens(_._1, (t, o) => (o, t._2))
+
+  /** Returns a scan that inputs/outputs pairs of elements, with `I` and `O` in the second element of the pair. */
+  def second[A]: Scan[S, (A, I), (A, O)] =
+    lens(_._2, (t, o) => (t._1, o))
+
+  /** Like [[lens]] but some elements are passed to the output (skipping the original scan) while other elements
+    * are lensed through the original scan.
+    */
+  def semilens[I2, O2](extract: I2 => Either[O2, I], inject: (I2, O) => O2): Scan[S, I2, O2] =
+    Scan[S, I2, O2](initial)(
+      (s, i2) =>
+        extract(i2).fold(
+          o2 => s -> Chunk.singleton(o2),
+          i => {
+            val (s2, os) = transform(s, i)
+            (s2, os.map(o => inject(i2, o)))
+          }
+        ),
+      _ => Chunk.empty
+    )
+
+  /** Like [[semilens]] but the elements of the original scan are output directly. */
+  def semipass[I2, O2 >: O](extract: I2 => Either[O2, I]): Scan[S, I2, O2] =
+    semilens(extract, (_, o) => o)
+
+  /** Returns a scan that wraps the inputs/outputs with `Either`.
+    * Elements on the left pass through the original scan while elements on
+    * the right pass through directly.
+    */
+  def left[A]: Scan[S, Either[I, A], Either[O, A]] =
+    semilens(_.fold(i => Right(i), a => Left(Right(a))), (_, o) => Left(o))
+
+  /** Returns a scan that wraps the inputs/outputs with `Either`.
+    * Elements on the right pass through the original scan while elements on
+    * the left pass through directly.
+    */
+  def right[A]: Scan[S, Either[A, I], Either[A, O]] =
+    semilens(_.fold(a => Left(Left(a)), i => Right(i)), (_, o) => Right(o))
+
+  /** Combines this scan with the supplied scan such that elements on the left
+    * are fed through this scan while elements on the right are fed through the
+    * suppplied scan. The outputs are joined together.
+    */
+  def or[S2, I2, O2 >: O](that: Scan[S2, I2, O2]): Scan[(S, S2), Either[I, I2], O2] =
```

Comment:
This function looks similar to the `choice` method from the `cats.arrow.Choice` typeclass, except that whereas the typeclass assumes the same `F`, here the state type would be part of that type.

Reply:
Renamed to `choice`